sw/device/lib/dif/dif_kmac.c

```diff
@@ -4,4 +4,478 @@
+  }
+
+  // Set key strength and calculate rate (r) and digest length (d) in bytes.
+  uint32_t kstrength;
+  switch (mode) {
+    case kDifKmacModeSha3Len224:
+      kstrength = KMAC_CFG_KSTRENGTH_VALUE_L224;
+      kmac->offset = 0;
+      kmac->r = calculate_rate_bits(224) / 8;
+      kmac->d = 224 / 8;
+      break;
+    case kDifKmacModeSha3Len256:
+      kstrength = KMAC_CFG_KSTRENGTH_VALUE_L256;
+      kmac->offset = 0;
+      kmac->r = calculate_rate_bits(256) / 8;
+      kmac->d = 256 / 8;
+      break;
+    case kDifKmacModeSha3Len384:
+      kstrength = KMAC_CFG_KSTRENGTH_VALUE_L384;
+      kmac->offset = 0;
+      kmac->r = calculate_rate_bits(384) / 8;
+      kmac->d = 384 / 8;
+      break;
+    case kDifKmacModeSha3Len512:
+      kstrength = KMAC_CFG_KSTRENGTH_VALUE_L512;
+      kmac->offset = 0;
+      kmac->r = calculate_rate_bits(512) / 8;
+      kmac->d = 512 / 8;
+      break;
+    default:
+      return kDifKmacBadArg;
+  }
+
+  // Hardware must be idle to start an operation.
+  if (!is_state_idle(kmac->params)) {
+    return kDifKmacError;
+  }
+
+  // Configure SHA-3 mode with the given strength.
+  uint32_t cfg_reg =
+      mmio_region_read32(kmac->params.base_addr, KMAC_CFG_REG_OFFSET);
+  cfg_reg =
+      bitfield_field32_write(cfg_reg, KMAC_CFG_KSTRENGTH_FIELD, kstrength);
+  cfg_reg = bitfield_field32_write(cfg_reg, KMAC_CFG_MODE_FIELD,
+                                   KMAC_CFG_MODE_VALUE_SHA3);
+  if (!checked_mmio_region_write32(kmac->params.base_addr, KMAC_CFG_REG_OFFSET,
+                                   cfg_reg)) {
+    // Write to configuration register failed.
+    return kDifKmacError;
+  }
+
+  // Issue start command.
+  uint32_t cmd_reg =
+      bitfield_field32_write(0, KMAC_CMD_CMD_FIELD, KMAC_CMD_CMD_VALUE_START);
+  mmio_region_write32(kmac->params.base_addr, KMAC_CMD_REG_OFFSET, cmd_reg);
+
+  // Poll until the status register is in the 'absorb' state.
+  while (true) {
+    if (is_state_absorb(kmac->params)) {
+      break;
+    }
+    // TODO(#6248): check for error.
+  }
+
+  return kDifKmacOk;
+}
+
+dif_kmac_result_t dif_kmac_absorb(dif_kmac_t *kmac, const void *msg, size_t len,
+                                  size_t *processed) {
+  // Set the number of bytes processed to 0.
+  if (processed != NULL) {
+    *processed = 0;
+  }
+
+  if (kmac == NULL || (msg == NULL && len != 0)) {
+    return kDifKmacBadArg;
+  }
+
+  // Check that an operation has been started.
+  if (kmac->r == 0) {
+    return kDifKmacError;
+  }
+
+  // Poll until the the status register is in the 'absorb' state.
+  if (!is_state_absorb(kmac->params)) {
+    return kDifKmacError;
+  }
+
+  // Copy the message one byte at a time.
+  // This could be sped up copying a word at a time but be careful
+  // about message endianness (e.g. only copy a word at a time when in
+  // little-endian mode).
+  for (size_t i = 0; i < len; ++i) {
+    mmio_region_write8(kmac->params.base_addr, KMAC_MSG_FIFO_REG_OFFSET,
```

Comment:
I've struggled a bit with this because the behaviour of big-endian mode is unclear when using an untyped `void *` data input. For example, if the data is unaligned do we send the first few bytes into the `MSG_FIFO` 1-by-1 until it is aligned? If so those bytes won't get reordered.

Big-endian mode only really makes sense to me when we have type information available for the data that we want to hash and can send it into the `MSG_FIFO` in chunks of the correct size.

We could return an error if data that is not aligned to word boundaries in big-endian mode. Would that be a reasonable solution?

Reply:
> 8B writes will potentially incur an additional penalty compared to 32B accesses.

I'm happy to do that optimization but I'd like to leave it for a more focused follow up PR. This one is already more complicated than I would like.

Reply:
SG.

Reply:
Disallowing arbitrary length unaligned messages would be an unusual decision for a hash library (for digests it makes more sense). While verification might not need it I think it is worth keeping for the "reference code" use case. In addition it keeps this code consistent with the HMAC DIF that does allow arbitrary length unaligned messages. Once I [optimize the code](https://github.com/lowRISC/opentitan/pull/6066#discussion_r622957881) the performance overhead for word aligned messages should be tiny.